app/controllers/notifications_controller.rb

```diff
@@ -26,12 +26,20 @@ def index
                        @user.notifications
                      end
 
-    @notifications = @notifications.includes(:notifiable).without_past_aggregations.order(notified_at: :desc)
+    @notifications = @notifications.without_past_aggregations.order(notified_at: :desc)
 
     # if offset based pagination is invoked by the frontend code, we filter out all earlier ones
     @notifications = @notifications.where("notified_at < ?", notified_at_offset) if notified_at_offset
 
-    @notifications = NotificationDecorator.decorate_collection(@notifications.limit(num))
+    @notifications = @notifications.limit(num)
+
+    # after notifications are paginated/limited, we check if any of the notifiables is "Comment" with no action
+    # because its respective view is the only one that actually uses ".notifiable" to render,
+    # this way we save useless eager loadings
+    notifiable_eager_loading_params = { notifiable_type: %w[Comment], action: [nil, ""] }
+    @notifications = @notifications.includes(:notifiable) if @notifications.exists?(notifiable_eager_loading_params)
```

Comment:
This is probably the most controversial change, because it does a `SELECT exists` for each pagination. The reason is that we were performing an eager load (which is a `SELECT *`) even when none of the rendered notifications rendered the "notifiable" object. Here we add a quick `SELECT exists` (which runs on 8 or 30 items), to potentially avoid including data we don't need to render.

Reply:
In giving this another look, the only situations we call `.notifiable` at all are in `_article.html.erb` and `_comment.html.erb`.

In `_article...` the code is...

```html
    <div>
      <% if json_data["article"]["published_at"] %>
        <small><%= time_ago_in_words json_data["article"]["published_at"] %> ago</small>
      <% else %>
        <small><%= time_ago_in_words notification.notifiable.published_at %> ago</small>
      <% end %>
    </div>
```

Which must be like this because we added "published at" to notifications later and some might not have this? This conditional can probably safely be removed in favor of the `json_data["article"]["published_at"]` by now as the old ones have faded away.

And then in `_comment...` we have...

```html
      <div>
        <small><%= time_ago_in_words notification.notifiable.created_at %> ago</small>
      </div>
```

Which seems similar to above in that the `json_data` must not include `created_at` so we have to call this, but I think the better solution would be to add this info to the `json_data` itself, and then add a conditional for its existence?

I think that conditional doesn't need to actually render the time if it doesn't exist. I think it could basically be like...

```html
  <% if json_data["comment"]["created_at"] %>
      <div>
        <small><%= time_ago_in_words json_data["comment"]["created_at"] %> ago</small>
      </div>
  <% end %>
```

And `created_at` needs to get added to the `comment_data`. As long as it doesn't mess up the formatting I don't think we need to render anything if `json_data["comment"]["created_at"]` doesn't exist unless it messes with formatting because this isn't super critical info and comment notifications are transient and will soon fade away. We could backfill info if that was needed, but I don't think it's necessary.

TL;DR — We do this, I don't think we need to call the `includes` at all. Seems like the includes only deals with a couple unnecessary exceptions to the general serialization approach.

This is explanation was pretty complicated, does it all make sense @rhymes?

Reply:
It makes perfect sense! I'll modify the article partial and add the date `comment_data`. Do you think it should be in this PR or in a following one?

Reply:
Why not add it to this PR? I'm also good with a followup since I know you will do it right away 😉

Reply:
I realized it's a bigger change than I anticipated. I have to change `app/services/notifications.rb`, probably the two workers related to sending these two notifications and likely a bunch of tests. I'd feel more comfortable doing this in two steps :D